### PR TITLE
Cache hashable input signature for _cache_key

### PR DESCRIPTION
**Pull request by jonathanchu33**
Evaluations of with-signature benchmarks pending merge of PR #41714

### CODE REVIEW

File: tensorflow/python/eager/function.py [hunk 1]
```diff
@@ -2940,6 +2942,11 @@ def flat_input_signature(self):
     """Returns the flattened input signature."""
     return self._function_spec.flat_input_signature
 
+  @property
+  def hashable_input_signature(self):
```
Comment (kkimdev) [inline tensorflow/python/eager/function.py:2946]: afaict, this abstraction doesn't provide much benefit and adds an unnecessary indirection perf cost.  Let's just use `self._hashable_input_signature` directly.

**Review by kkimdev — COMMENTED** [review]
- Let's land your benchmark PR first.
- Please include the perf difference before&after in the PR message.

File: tensorflow/python/eager/function.py [hunk 2]
```diff
@@ -2902,6 +2902,8 @@ def __init__(self,
     self._function_attributes = attributes or {}
     self._capture_by_value = capture_by_value
     self.tracing_count = 0
+    self._hashable_input_signature = _make_input_signature_hashable(
+        self.flat_input_signature)
```
Comment (sun51) [inline tensorflow/python/eager/function.py:2906]: we should only do this when the self.input_signature is not None

File: tensorflow/python/eager/function.py [hunk 3]
```diff
@@ -2940,6 +2942,11 @@ def flat_input_signature(self):
     """Returns the flattened input signature."""
     return self._function_spec.flat_input_signature
 
+  @property
```
Comment (sun51) [inline tensorflow/python/eager/function.py:2945]: agree, let us remove this method, directly use self._hashable_input_signature

File: tensorflow/python/eager/function.py [hunk 4]
```diff
@@ -3072,10 +3075,12 @@ def _cache_key(self, args, kwargs, include_tensor_ranks_only=False):
       inputs = (args, kwargs) if kwargs else args
       input_signature = pywrap_tfe.TFE_Py_EncodeArg(inputs,
                                                     include_tensor_ranks_only)
+      hashable_input_signature = _make_input_signature_hashable(input_signature)
     else:
       del args, kwargs
       assert not include_tensor_ranks_only
-      input_signature = self.flat_input_signature
+      assert hasattr(self, '_hashable_input_signature')
```
Comment (sun51) [inline tensorflow/python/eager/function.py:3082]: why is this needed?

**Review by sun51 — COMMENTED** [review]
overall LGTM, thanks, just one minor comment

**Comment by jonathanchu33** [comment]
About a 10us speedup from roughly 190us to 180us, observed as an approximate median over 5-10 trials with 100000 iterations each on benchmark_defun_matmul_2_by_2_CPU_with_signature

File: tensorflow/python/eager/function.py (hunk 1 above)
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py]: Done

File: tensorflow/python/eager/function.py (hunk 2 above)
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py]: Done

File: tensorflow/python/eager/function.py (hunk 3 above)
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py]: Done

File: tensorflow/python/eager/function.py (hunk 4 above)
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py:3082]: Just to make sure the variable was initialized at object creation, since it only happens in the `input_signature is not None` case. In theory, I don't think the input_signature attribute changes after a Function object is initialized, and it doesn't look like we are doing any error handling if asserts fail, so I can remove if it feels extraneous. Thoughts?

Comment (sun51) [reply tensorflow/python/eager/function.py:3082]: i think you can remove it, it is extra.

Comment (jonathanchu33) [reply tensorflow/python/eager/function.py]: Done

**Review by sun51 — DISMISSED** [review]

File: tensorflow/python/eager/function.py [hunk 5]
```diff
@@ -3144,9 +3148,9 @@ def _cache_key(self, args, kwargs, include_tensor_ranks_only=False):
           save_context.get_save_options().experimental_variable_policy)
     else:
       variable_policy = save_options.VariablePolicy.EXPAND_DISTRIBUTED_VARIABLES
-
+
```
Comment (kkimdev) [inline tensorflow/python/eager/function.py:3151]: nit: remove trailing white space

**Review by kkimdev — DISMISSED** [review]

File: tensorflow/python/eager/function.py (hunk 5 above)
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py]: Done

**Review by kkimdev — APPROVED** [review]

**Review by sun51 — APPROVED** [review]
